### DISCUSSION TITLE
72. 데이터베이스가 여러 인덱스 중 하나를 선택하는 기준에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#74] [General]
데이터베이스가 여러 인덱스 중 하나를 선택하는 기준에 대해 설명해 주세요.

**Comment by eun-byeol** [comment]
MySQL의 비용 기반 최적화 방식의 경우 `옵티마이저`가 `인덱스 통계 정보를 이용`해 사용할 인덱스를 결정합니다.

쿼리 처리를 위한 여러 가지 방법을 만들고, 비용을 산출하여 최소 비용의 처리 방식을 선택합니다.

일반적으로 인덱스의 `선택도`(Selectivity)가 높고, `쿼리 조건절`(WHERE, JOIN 등)에 포함된 칼럼이 인덱스와 맞는 경우 우선순위가 높습니다.

**Comment by eun-byeol** [comment]
## 옵티마이저 종류

> 비용 기반 최적화(Cost-based optimizer)

- 현재 대부분의 DBMS에서 선택하는 방식
- 쿼리 처리를 위한 여러 가지 방법을 만들고, 각 단위 작업의 비용 정보와 대상 테이블의 예측된 통계 정보를 이용해 실행 계획별 비용 산출
- 산출된 실행 방법별 최소 비용의 처리 방식을 선택

> 규칙 기반 최적화 방법(Rule-based optimizer)
 
- 대상 테이블의 레코드 건수나 선택도 등을 고려하지 않고 옵티마이저에 내장된 우선순위에 따라 실행 계획을 수립하는 방식
- 통계 정보를 사용하지 않음
- 같은 쿼리에 대해 거의 항상 같은 실행 방법을 선택

## 옵티마이저가 풀 테이블 스캔을 선택하는 경우

- 테이블 레코드 건수가 너무 작아 인덱스를 통해 읽는 것보다 풀 테이블 스캔이 더 빠른 경우
- WHERE 절이나 ON 절에 인덱스를 이용할 수 있는 적절한 조건이 없는 경우
- 인덱스 레인지 스캔을 사용할 수 있는 쿼리라도 옵티마이저가 판단한 조건 일치 레코드 건수가 너무 많은 경우
